pallets/vector/src/lib.rs

```diff
@@ -395,72 +410,188 @@ pub mod pallet {
 			Weight::zero()
 		}
 	}
+	// TODO: Rename to FunctionInput
+	#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
+	pub struct FunctionInputs {
+		pub updates: Vec<Update>,
+		pub finality_update: FinalityUpdate,
+		pub expected_current_slot: u64,
+		pub store: LightClientStore,
+		pub genesis_root: Bytes32,
+		pub forks: Forks,
+		pub execution_state_proof: ExecutionStateProof,
+	}
 
+	#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
+	pub struct ExecutionStateProof {
+		#[serde(rename = "executionStateRoot")]
+		pub execution_state_root: B256,
+		#[serde(rename = "executionStateBranch")]
+		pub execution_state_branch: Vec<B256>,
+		pub gindex: String,
+	}
+	/// Merkle branch index & depth for the execution state root proof.
+	pub const MERKLE_BRANCH_INDEX: usize = 802;
+	pub const MERKLE_BRANCH_DEPTH: usize = 9;
 	#[pallet::call]
 	impl<T: Config> Pallet<T>
 	where
 		[u8; 32]: From<T::AccountId>,
 	{
 		/// The entrypoint for fulfilling a call.
 		/// function_id Function identifier.
-		/// input Function input.
-		/// output Function output.
-		/// proof  Function proof.
-		/// slot  Function slot to update.
+		/// inputs Function input.
 		#[pallet::call_index(0)]
+		// Unused parameter that's hard to remove, would mess up traits
 		#[pallet::weight(weight_helper::fulfill_call::<T>(* function_id))]
 		pub fn fulfill_call(
 			origin: OriginFor<T>,
 			function_id: H256,
-			input: FunctionInput,
-			output: FunctionOutput,
-			proof: FunctionProof,
-			#[pallet::compact] slot: u64,
+			inputs: Vec<u8>, // TODO: Convert to fixed bytes
 		) -> DispatchResultWithPostInfo {
+			let config = ConfigurationStorage::<T>::get();
+			let FunctionInputs {
+				updates,
+				finality_update,
+				expected_current_slot,
+				mut store,
+				genesis_root,
+				forks,
+				execution_state_proof,
+			} = serde_cbor::from_slice(&inputs).unwrap();
+
+			let mut is_valid = true;
+			let prev_head = store.finalized_header.slot;
+
+			// 1. Apply sync committee updates, if any
+			for (index, update) in updates.iter().enumerate() {
+				is_valid = is_valid
+					&& verify_update(
+					update,
+					expected_current_slot,
+					&store,
+					genesis_root.clone(),
+					&forks,
+				)
+					.is_ok();
+
+				apply_update(&mut store, update);
```

Comment:
That will apply the `update` even if it was invalid. 
Do we want this behavior? Should not we raise up an error in that case (short-circuit)?